crates/dbsp/src/trace/spine_async/mod.rs

```diff
@@ -83,81 +76,369 @@ impl<B: Batch + Send + Sync> From<&Spine<B>> for CommittedSpine<B> {
     }
 }
 
-/// A unique identifier for batches we hold within levels.
-#[derive(SizeOf, Ord, Default, PartialOrd, Debug, Copy, Clone, Eq, PartialEq)]
-struct BatchIdent {
-    /// The number of nested batches, either 1 if no merging is going on or 2 if
-    /// this represents a pair of batches being merged. (Larger values would
-    /// represent multiway merges, which we currently don't do.)
-    #[size_of(skip)]
-    n: usize,
-    /// Which level this batch is from.
+/// A group of batches with similar sizes (as determined by [size_from_level]).
+struct Slot<B>
+where
+    B: Batch,
+{
+    /// Optionally, a pair of batches that are currently being merged.  These
+    /// batches are not in `loose_batches`.
     ///
-    /// We use this to find the two batches we want to remove when
-    /// we finished a merge.
-    level: usize,
-    /// A key that is unique within the level.
+    /// Invariant: the batches (if present) must be non-empty.
+    merging_batches: Option<[Arc<B>; 2]>,
+
+    /// Zero or more batches not currently being merged.
     ///
-    /// We use this to find the two batches we want to remove when
-    /// we finished a merge.
-    #[size_of(skip)]
-    key: u64,
+    /// Invariant: the batches must be non-empty.
+    loose_batches: Vec<Arc<B>>,
 }
 
-/// Describes the state of a layer.
-///
-/// A layer can be empty, contain a single batch, or contain multiple batches
-/// (currently just 2) that are being merged in the background.
+impl<B> Default for Slot<B>
+where
+    B: Batch,
+{
+    fn default() -> Self {
+        Self {
+            merging_batches: None,
+            loose_batches: Vec::new(),
+        }
+    }
+}
+
+impl<B> Slot<B>
+where
+    B: Batch,
+{
+    /// If this slot doesn't currently have an ongoing merge, and it does have
+    /// at least two loose batches, picks two of the loose batches and makes
+    /// them into merging batches, and returns those batches. Otherwise, returns
+    /// `None` without changing anything.
+    fn try_start_merge(&mut self) -> Option<[Arc<B>; 2]> {
```

Comment:
Since batches in a slot can differ in size up to 10x, should we put some extra effort to merge similar-size batches?

Reply:
Some possible merge policies:
* If we always merge the smallest batches of size k in a slot, we will pile up batches of size 2k in that slot if new batches of size k keep getting added.
* If we always merge the largest batches, that's inefficient.
* We currently always merge the batches most recently added. The merge result (if it goes to the same slot) becomes the most recently added. So we'll end doing merges of k*n + k => k*(n+1)  until that overfills the slot, if we keep getting new batches of size k. Not ideal either.
* Another policy would be to merge the two least recently added batches. Then we'll do k+k=>2k, ..., k+k=>2k, 2k+2k=>4k, ..., 2k+2k=>4k, 4k+4k=>8k, ... and so on. It could also be good for GC to work with older batches (as you observed). That might be a good policy. (It could be bad for cache locality, since we're working with the oldest data.)
* Another variation would be to merge the least recently added batch with the other batch closest in size. I don't have an intuition about this.

Reply:
I filed a new issue about merge policies: #2124